### PR TITLE
sp_DatabaseRestore: harden against parameter SQL injection

### PR DESCRIPTION
**Pull request by BrentOzar**
## Summary

Hardens `sp_DatabaseRestore` against parameter-based SQL injection. Threat model: a caller with `EXECUTE` on the proc but **not** sysadmin should not be able to escalate to arbitrary T-SQL or shell command execution by passing crafted parameter values.

- **Validation gate** for path-shaped parameters. Rejects `"`, `&`, `|`, `;`, `^`, `<`, `>`, NUL, CR, LF — characters with no legitimate use in Windows paths but high command-injection risk. Single quotes are allowed (legal but rare in real paths) and escaped at concat sites instead. Applied to `@BackupPathFull/Diff/Log`, `@MoveDataDrive/LogDrive/FilestreamDrive/FullTextCatalogDrive`, `@StandbyUndoPath`, `@FileNamePrefix`. (`@Database` is **not** in this list — it's a database identifier, not a path; `&`/`;` are legal in DB names. It's protected via single-quote escaping at the few sites it gets concatenated into a quoted SQL literal.) The LIKE pattern uses `COLLATE Latin1_General_BIN2` so the `[…]` class catches NCHAR(0) — under the default collation NUL is silently dropped from both pattern and value.
- **`QUOTENAME(@DatabaseOwner)`** in the `ALTER AUTHORIZATION` build (was `'[' + @DatabaseOwner + ']'`, which a `]` in the value would escape).
- **`PARSENAME` + `QUOTENAME` for `@RunStoredProcAfterRestore`** before the `EXEC` build. Rejects 3- or 4-part names; quotes both the schema and proc parts. Always emits a 3-part name — for 1-part input the schema slot is left empty (`[db]..[proc]`) so SQL Server's own schema resolution kicks in; without the second dot, `[db].[proc]` parses as schema.object in the *current* DB. Closes the highest-severity site — previously the value was concatenated raw and passed straight to `sp_executesql @sql`.
- **Single-quote escape at every dynamic-RESTORE concat site** that wraps a caller-supplied path or `@Database` in `'...'`. Includes single-file `RESTORE DATABASE/LOG FROM DISK = '...'`, `STANDBY = '...'`, and the per-row split-backup `DISK=` entries built via `FOR XML PATH`. Note: this is **not** applied to the `xp_cmdshell DIR /b "..."` command-string build — cmd.exe doesn't interpret `''` as an escape, so doubling there would corrupt legitimate paths with an apostrophe; the value is just passed as-is to xp_cmdshell and never crosses a SQL parser again.
- **`, TYPE).value('.', 'nvarchar(max)')`** on the three `FOR XML PATH` split-backup builders so legitimate filenames containing `&`/`<`/`>` don't get XML-entity-encoded into the generated `RESTORE DISK=` paths.
- **Doubled-twice escape (`REPLICATE(N'''', 4)`) for `{Path}` substitution** in the `RESTORE HEADERONLY` / `RESTORE FILELISTONLY` templates, where the path lives inside a nested `EXEC` literal (two parser layers).

Out of scope (deliberate): filenames sourced from the filesystem (xp_dirtree / xp_cmdshell output) and `RESTORE FILELISTONLY` output flowing back into MOVE clauses. An attacker who can write into the backup folder is already trusted in this model. (The XML-PATH split-backup sites do escape `BackupFile` alongside `BackupPath` for code-shape consistency, but that's a side effect, not the intended fix.)

## Test plan

Verified end-to-end against SQL Server 2025 RTM-CU4. All eleven injection probes plus regression cases pass:

- [x] Deploy + verify the proc compiles cleanly.
- [x] Regression — legitimate paths still work: `C:\Backups\`, `\\server\share\`, and a path containing a single quote (`C:\Backups\It's Friday\`). The xp_cmdshell `DIR /b` command preserves the apostrophe correctly (single quote, not doubled).
- [x] `@DatabaseOwner = 'sa'` and `@RunStoredProcAfterRestore = 'dbo.MyProc'` (and `'MyProc'`) — both accepted; the generated `ALTER AUTHORIZATION` and `EXEC` strings (visible with `@Debug = 1, @Execute = 'N'`) show `[sa]` and `[Test].[dbo].[MyProc]` / `[Test]..[MyProc]` respectively.
- [x] `@Database = 'My&DB'` — passes validation (database identifier, not a path).
- [x] Injection attempts blocked or neutralized:
  - `@BackupPathFull = 'C:\Backups\"& whoami &"'` — validation gate rejects with RAISERROR + RETURN.
  - `@BackupPathFull = 'C:\Backups\; xp_cmdshell ''calc''; --'` — same, blocked by `;` rejection.
  - `@MoveDataDrive` containing NCHAR(0) — caught by the BIN2-collated LIKE.
  - `@DatabaseOwner = 'sa]; DROP TABLE foo; --'` — `QUOTENAME` produces a single bracketed identifier; the `EXISTS` check in `master.dbo.syslogins` fails to find that login and the proc falls through to the "not a valid Login" PRINT path with no injected statement executed.
  - `@RunStoredProcAfterRestore = 'dbo.MyProc; DROP TABLE foo; --'` — `PARSENAME(..., 1)` returns `MyProc; DROP TABLE foo; --`, gets wrapped in brackets; the `EXEC` fails to find a procedure with that bracketed name.
  - `@RunStoredProcAfterRestore = 'srv.db.dbo.MyProc'` — rejected with the new RAISERROR (4-part name not allowed).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: sp_DatabaseRestore.sql [hunk 1]
```diff
@@ -1672,8 +1696,23 @@ END;'
...
+	   OR PARSENAME(@RunStoredProcAfterRestore, 4) IS NOT NULL
+	BEGIN
+		RAISERROR('@RunStoredProcAfterRestore must be a procedure name or schema.procedure (1- or 2-part name).', 16, 1) WITH NOWAIT;
+		RETURN;
+	END;
 	PRINT 'Attempting to run ' + @RunStoredProcAfterRestore
-	SET @sql = N'EXEC ' + @RestoreDatabaseName + '.' + @RunStoredProcAfterRestore
+	/* Always emit a 3-part name (db.schema.proc). For 1-part input the schema slot is left empty
+	   so the name resolves as db..proc — i.e., the default schema in the restored DB. Without the
+	   second dot, [db].[proc] is parsed as schema.object in the *current* DB. */
```
[the remark this answers is not in the document]
Comment (BrentOzar) [reply sp_DatabaseRestore.sql]: This isn't quite right — `db..object` does **not** uniformly resolve to `db.dbo.object`. SQL Server applies the caller's default schema in the target database first, falling back to `dbo` only if no object exists under that schema.

Verified empirically on SQL Server 2025 (RTM-CU4):

```sql
USE TestRR;
EXEC ('CREATE SCHEMA myschema');
CREATE PROCEDURE dbo.SchemaTest      AS PRINT 'I am dbo.SchemaTest';
CREATE PROCEDURE myschema.SchemaTest AS PRINT 'I am myschema.SchemaTest';
CREATE USER TestUser WITHOUT LOGIN WITH DEFAULT_SCHEMA = myschema;
GRANT EXECUTE ON dbo.SchemaTest      TO TestUser;
GRANT EXECUTE ON myschema.SchemaTest TO TestUser;

EXECUTE AS USER = 'TestUser';
EXEC TestRR..SchemaTest;        -- prints "I am myschema.SchemaTest"
REVERT;
```

So `db..proc` is the right pattern when we want SQL Server's normal schema-resolution rules to apply in the restored DB. I've kept the code as-is and just retightened the comment to "SQL Server applies its own schema-resolution rules in the target DB" so it sidesteps the dbo-vs-default debate.

File: sp_DatabaseRestore.sql [hunk 2]
```diff
@@ -1672,8 +1696,23 @@ END;'
...
+	DECLARE @RunStoredProcSchema sysname = NULLIF(PARSENAME(@RunStoredProcAfterRestore, 2), N'');
+	DECLARE @RunStoredProcName   sysname = PARSENAME(@RunStoredProcAfterRestore, 1);
+	IF @RunStoredProcName IS NULL
+	   OR PARSENAME(@RunStoredProcAfterRestore, 3) IS NOT NULL
+	   OR PARSENAME(@RunStoredProcAfterRestore, 4) IS NOT NULL
+	BEGIN
+		RAISERROR('@RunStoredProcAfterRestore must be a procedure name or schema.procedure (1- or 2-part name).', 16, 1) WITH NOWAIT;
+		RETURN;
+	END;
 	PRINT 'Attempting to run ' + @RunStoredProcAfterRestore
```
[the remark this answers is not in the document]
Comment (BrentOzar) [reply sp_DatabaseRestore.sql]: Good catch — pushed in c1bb7cc. The fix LTRIMs/RTRIMs the input and then collapses any whitespace adjacent to a dot before `PARSENAME`, so `'dbo. MyProc'`, `'  MyProc  '`, and `'[dbo]. [My Proc]'` all parse the same as the no-whitespace form.

One caveat I want to flag: this isn't a full identifier parser. If someone passes a bracketed identifier whose own characters include a literal dot followed by whitespace — e.g. `'[my . proc]'` — the normalization will eat the surrounding whitespace and turn it into `'[my.proc]'`, which would then mis-parse. Procedure names of that shape are vanishingly rare, so I think the trade-off is worth it, but happy to swap in a bracket-aware parser if you'd rather close that gap.

[the remark this answers is not in the document]
Comment (BrentOzar) [reply sp_DatabaseRestore.sql]: Verified end-to-end on SQL Server 2025 with all three problematic shapes — `'  MyProc  '`, `'dbo. MyProc'`, `'[dbo]. [My Proc]'` — and they all pass the validation gate now and reach the proc's normal "no rows in path" failure (i.e. exactly like the no-whitespace forms). Test cases 3.6–3.8 in `Documentation/Development/Test sp_DatabaseRestore.sql` cover them. Resolving — the bracketed-name-with-internal-dot edge case I flagged remains a known trade-off, captured in the inline comment.
